samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -541,57 +594,23 @@ int consume_device_message(void)
 		send_failure_count = 0;
 	}
 
-	/* Clean up. */
-	k_free(msg);
-
 	return ret;
 }
 
-int send_device_message(const char *const msg)
-{
-	/* Copy message data onto the heap (will be freed by the message consumer when done). */
-	size_t len = strlen(msg) + 1;
-	char *msg_buf = k_malloc(len);
-
-	if (!msg_buf) {
-		LOG_ERR("Could not alloc memory for new device message");
-		return -ENOMEM;
-	}
-	strncpy(msg_buf, msg, len);
-	LOG_DBG("Enqueued message: %s", msg_buf);
-
-	/* Attempt to append data onto message queue. */
-	if (k_msgq_put(&device_message_queue, &msg_buf, K_NO_WAIT)) {
-		LOG_ERR("Device message rejected, outgoing message queue is full");
-		k_free(msg_buf);
-		return -ENOMEM;
-	}
-
-	return 0;
-}
-
-int send_device_message_cJSON(cJSON *msg_obj)
+int send_device_message(struct nrf_cloud_obj *const msg_obj)
 {
-	int ret = 0;
-	char *msg = NULL;
+	int ret = enqueue_device_message(msg_obj, true);
 
-	if (!msg_obj) {
-		LOG_ERR("Cannot send NULL device message object");
-		return -EINVAL;
-	}
-
-	/* Convert message object to a string. */
-	msg = cJSON_PrintUnformatted(msg_obj);
-	if (msg == NULL) {
-		LOG_ERR("Failed to convert cJSON device message object to string");
-		return -ENOMEM;
+	if (ret) {
+		LOG_ERR("Cannot add message to queue");
+		nrf_cloud_obj_free(msg_obj);
+	} else {
+		/* The message data now belongs to the queue.
+		 * Reset the provided object so it cannot be modified.
+		 */
+		nrf_cloud_obj_reset(msg_obj);
```

Comment:
I guess this is the use case I was missing in my earlier comment.

But if the user is handing off ownership of the object to `connection.c`, enforcing that with `nrf_cloud_obj_reset` seems unnecessary to me, so `nrf_cloud_obj_reset` could still maybe be removed

Reply:
`nrf_cloud_obj_reset` is really just a convenience function for such a case.
the user could directly set the pointers to NULL if they want.